optimism-children-indexer/optimism-children-indexer-logic/src/services/layer3_indexer/indexer.rs

```diff
@@ -0,0 +1,278 @@
+};
+
+///  TODO: Consider making this configurable via Settings
+/// Time between database chains refreshes
+const REFRESH_CHAINS_INTERVAL: Duration = Duration::from_secs(15);
+/// Restart delay for fully synced chains
+const RESTART_DELAY_SYNCED: Duration = Duration::from_mins(5);
+/// Restart delay for chains falling behind
+const RESTART_DELAY_BEHIND: Duration = Duration::from_secs(5);
+/// Restart delay for failing chains
+const RESTART_DELAY_FAILING: Duration = Duration::from_secs(90);
+
+/// Main task for Layer3 Indexer.
+pub struct Layer3Indexer {
+    db: Arc<DatabaseConnection>,
+    chains: HashMap<i64, Layer3Chain>,
+    tasks: JoinSet<(i64, Result<Layer3IndexerTaskOutput>)>,
+    abort_handles: HashMap<i64, AbortHandle>,
+}
+
+impl Layer3Indexer {
+    pub fn new(db: Arc<DatabaseConnection>) -> Self {
+        Self {
+            db,
+            chains: HashMap::new(),
+            tasks: JoinSet::new(),
+            abort_handles: HashMap::new(),
+        }
+    }
+
+    /// Fetch enabled chains from database
+    async fn refresh_chains(&mut self) -> Result<()> {
+        self.chains = optimism_children_l3_chains::Entity::find()
+            .all(&*self.db)
+            .await?
+            .into_iter()
+            .filter(|config| config.enabled)
+            .map(|config| (config.chain_id, config))
+            .collect();
+
+        Ok(())
+    }
+
+    /// Synchronize running tasks with current chain list
+    /// Spawns tasks for new chains, cancels tasks for removed/disabled chains
+    fn sync_tasks(&mut self) {
+        // Collect chains to spawn
+        let to_spawn: Vec<_> = self
+            .chains
+            .iter()
+            .filter(|(&chain_id, _)| !self.abort_handles.contains_key(&chain_id))
+            .map(|(_, chain)| chain.clone())
+            .collect();
+
+        // Spawn new tasks
+        for config in to_spawn {
+            tracing::debug!(
+                chain_id = config.chain_id,
+                "[{}] Spawning indexer task.",
+                config.chain_name
+            );
+
+            self.spawn_task(config, Duration::ZERO);
+        }
+
+        // Cancel tasks for chains that were removed or disabled
+        self.abort_handles.retain(|&chain_id, handle| {
+            if !self.chains.contains_key(&chain_id) {
+                tracing::debug!(
+                    chain_id = chain_id,
+                    "Cancelling task for removed/disabled chain."
+                );
+                handle.abort();
+                false
+            } else {
+                true
+            }
+        });
+    }
+
+    /// Spawn a new indexer task with delay
+    fn spawn_task(&mut self, config: Layer3Chain, delay: Duration) {
+        let chain_id = config.chain_id;
+
+        let handle = self.tasks.spawn(async move {
+            let chain_id = config.chain_id;
+            let task = Layer3IndexerTask::new(config);
+            let result = task.run_with_delay(delay).await;
```

Comment:
I don't have a good argument against how it's done here, but putting `sleep` in the restart logic instead of the run seems way more natural to me.